GeoParquet doc (#25)
developmentseed/cloud-optimized-geospatial-formats-guide: geoparquet/index.qmd

---
title: GeoParquet
subtitle: Guidelines for GeoParquet
---

# GeoParquet

GeoParquet is an encoding for how to store geospatial vector data (point, lines, polygons) in [Apache Parquet](https://parquet.apache.org/), a popular columnar storage format for tabular data.

> Parquet has a wide ecosystem of tools and support; GeoParquet builds on this existing success by defining how to store geometries in the Parquet format. Because GeoParquet is not a separate format, any program that can read Parquet is able to load GeoParquet as well, even if it can't make sense of the geometry information. This is very similar to how GeoTIFF layers geospatial information on top of the existing TIFF image standard.

Reviewer: ```suggestion
Parquet has a wide ecosystem of tools and support; GeoParquet builds on this success by defining how to store geometries in the Parquet format. Because GeoParquet is not a separate format, any program that can read Parquet is able to load GeoParquet as well, even if it can't make sense of the geometry information. This is very similar to how GeoTIFF layers geospatial information on top of the existing TIFF image standard.
```

> The two main things that GeoParquet defines on top of Parquet are how to encode geometries in the geometry column and how to include metadata like the geometries' CRS.

Reviewer 2: Not everyone knows what CRS means, should we link to a page or spell it out.

> GeoParquet is a relatively young format, and the specification has not yet reached a 1.0 release (as of August 2023, it's at [1.0.0-rc.1](https://github.com/opengeospatial/geoparquet/releases/tag/v1.0.0-rc.1)). However, reading and writing GeoParquet has been [supported in GDAL since version 3.5](https://gdal.org/drivers/vector/parquet.html), and thus can be used in programs like GeoPandas and QGIS.

Reviewer 3: Small note that GeoPandas 0.13.2's [`read_parquet`](https://geopandas.org/en/v0.13.2/docs/reference/api/geopandas.read_parquet.html) function doesn't actually go through GDAL but `pyarrow`. Using [`read_file(..., engine="pyogrio")`](https://geopandas.org/en/v0.13.2/docs/reference/api/geopandas.read_file.html) would go through GDAL though, but is apparently much slower than through pyarrow, see benchmarks at https://github.com/geopandas/geopandas/issues/2429#issuecomment-1126077276.

Reviewer 2: That should be a callout box, particularly in comparison to other formats (all other ogr formats we recommend pyogrio pretty much all the time right now)

Author: > Small note that GeoPandas 0.13.2's [`read_parquet`](https://geopandas.org/en/v0.13.2/docs/reference/api/geopandas.read_parquet.html) function doesn't actually go through GDAL but `pyarrow`.

Yes indeed, and this is important to call out.

> Using [`read_file(..., engine="pyogrio")`](https://geopandas.org/en/v0.13.2/docs/reference/api/geopandas.read_file.html) would go through GDAL though, but is apparently much slower than through pyarrow, see benchmarks at [geopandas/geopandas#2429 (comment)](https://github.com/geopandas/geopandas/issues/2429#issuecomment-1126077276).

The reason this is slower is because the default (original) pyogrio implementation calls GDAL's row-based `GetNextFeature` API, and then on the pyogrio side (in Cython) assembles columnar arrays. For most source datasets that are row based, whether the loop is in GDAL itself or in Cython doesn't matter. But here it does matter! Because effectively GDAL is transposing from column-oriented to row-oriented, and then pyogrio is doing the reverse!

If you use pyogrio's [Arrow interface](https://pyogrio.readthedocs.io/en/latest/api.html#pyogrio.raw.read_arrow), which implements [GDAL's RFC 86](https://gdal.org/development/rfc/rfc86_column_oriented_api.html) then it would likely be much faster, because there's no transpose back and forth. _But_ this is likely still significantly slower than using `pyarrow` directly because GDAL's RFC 86 implementation uses a fixed row group size (currently set to 65536 by default). So if this row group size doesn't match the underlying row group size in the Parquet files, GDAL will have to slice the original row group into the new size.

All of this is kinda moot anyways because the GDAL binary that is built as part of the pyogrio wheels don't include `libarrow`. So unless you build pyogrio from source, trying to load a GeoParquet file will give you "unknown data format".

Author: Thoughts on whether I should put the above explanation anywhere? It's probably more technical than we want to put into these docs

Because GeoParquet stores geometries in standard [Well-Known Binary](https://en.wikipedia.org/wiki/Well-known_text_representation_of_geometry#Well-known_binary) (WKB), it supports any vector geometry type defined in the OGC Simple Features specification. This includes the standard building blocks of Point, LineString, Polygon, MultiPoint, MultiLineString, MultiPolygon, and GeometryCollection. A best practice is to store only geometries with the same type, as that allows readers to know which geometry type is stored without scanning the entire file.

Some of the sections below will discuss strengths of Parquet in general. Keep in mind that because GeoParquet is built on top of Parquet, GeoParquet inherits all of these strengths.

> ### File layout

Reviewer 3: Use h2 instead of h3 here?

```suggestion
## File layout
```

Author: Yeah, we need to be consistent across documents. In the flatgeobuf doc I used h3 instead of h2. I should see what the others are using for their headings

Reviewer: Good to be consistent, I think keep h1 for the page title, h2 for sub-page section headers and h3 for any headers that are for sub-section content (like `Column oriented` below.

A Parquet file includes metadata describing the internal chunking. This metadata includes the byte range of every column chunk in the dataset. This allows a Parquet reader to fetch _any given column chunk_ once they have the file metadata.

The Parquet metadata also includes _column statistics_ (the minimum and maximum value) for each column chunk. This means that if a user is interested in data where column "A" has values greater than 100, the Parquet reader can skip loading and parsing any column chunks where the maximum is known to be less than 100.

In Parquet, the metadata is located at the _end_ of the file rather than at the beginning. This makes it much easier to write, as you don't need to know how many total rows you have at the beginning, but makes it slightly harder to read. In practice, this is not too much more difficult to read: a Parquet reader first reads the end of the file, then makes reads for select columns.

> ### Column oriented

Reviewer: Do you think column oriented should be hyphenated here and elsewhere in this document?

The bytes of each _column_ are contiguous, instead of each _row_. This means that it's easy to filter on columns — fetching _all rows of a single column_ — but not possible to filter on individual rows.

### Column filtering

Because Parquet is column oriented, a Parquet reader can fetch only specific columns that the user is interested in.

### Row group filtering

Because Parquet is internally chunked, Parquet can fetch only specific row groups that meet a specific filtering condition.

> Note that row group filtering on a specific column tends to only work well if the Parquet file was sorted on that column when saved. Non-sorted columns tend to have random values, and so the column statistics won't tend to filter out many row groups.

Reviewer: If I understand this correctly, does this mean that you can really only optimize filtering row groups by one column?

Author: Yeah, pretty much. I think that's the biggest difference between file formats and databases. Databases can have an index on whatever column you want, and then when you run a query, it'll use all of the indexes. But that's why it's hard to make databases work as cloud-native files, because if you have high latency, you don't want to make lots of tiny fetches.

FlatGeobuf optimizes filtering by a spatial query (when an index is generated) because it sorts all features in the file by their hilbert value. Parquet lets you sort by any column you choose. You _can_ sort by multiple columns, but the effectiveness of the sort drops off substantially after the first column.

Author: Do you think I should try to be more specific here? Or is this a general "cloud-native" note to go on a different page?

Reviewer: Good question. I think for now I would add

```
::: {.callout-note}
you can really only optimize filtering row groups by one column. This is the biggest difference between file formats and databases. Databases can have multiple indexes on whatever columns you want, and then when you run a query, and it will use all of the indexes. But that's why it's hard to make databases work as cloud-native files, because if you have high latency, you don't want to make lots of tiny fetches.
```

as a note. And then if we have a general cloud-native concepts page which explains more, we can link to that more general page.

I do think a page on databases vs files would be good, as I mentioned before, using some of this content and content from your DS presentation.

### Internal compression by default

> Parquet is internally compressed _by default_. Additionally, Parquet compression tends to be very efficient, leading to much smaller file sizes than comparable formats.

Reviewer: ```suggestion
Parquet is internally compressed _by default_ and Parquet compression is more efficient compared to other formats.

Compression algorithms are more effective when nearby bytes are more similar to each other. Data within a column tends to be much more similar than data across a row. Since Parquet is column oriented, compression algorithms work better and result in smaller file sizes than comparable row-based formats. As a result, Parquet file sizes are relatively smaller.
```

I find this structure slightly easier to follow

Reviewer: Fine as-is too, I find this slightly more readable.

Compression algorithms are more effective when nearby bytes are more similar to each other. Data within a column tends to be much more similar than data across a row. Since Parquet is column oriented, compression algorithms work better and result in smaller file sizes than on a comparable row-based format.

It's possible to have random access to one of the internal chunks inside the file at large, even though that chunk is compressed. Note that it isn't possible to fetch partial data _inside one chunk_ without loading and decompressing the entire chunk.

### Geometries encoded as Well-Known Binary

For maximum compatibility with existing systems, geometries are stored as [ISO-standard WKB](https://github.com/opengeospatial/geoparquet/blob/main/format-specs/geoparquet.md#encoding). Most geospatial programs are able to read and write WKB.

### No spatial index (yet!)

GeoParquet is a young specification, and spatial indices are not yet part of the standard. Future revisions of GeoParquet are expected to add support for spatial indexes.

> One way around this is to store multiple GeoParquet files according to some region identifier, cataloging each file with STAC.

Reviewer: ```suggestion
One way around this is to store multiple GeoParquet files according to some region identifier, cataloging each file with the [SpatioTemporal Asset Catalog (STAC) specification](https://stacspec.org).
```

### Multi-file support

While at medium data sizes GeoParquet is most easily distributed as a single file, at large data sizes a single dataset is often split into multiple files. Sometimes multiple files can be easier to write, such as if the data is output from a distributed system.

### Multithreading support but no streaming

In a streaming download, you read bytes starting at the beginning of the file, progressing towards the end. In Parquet, this is not helpful because the metadata is in the _footer_ of the file instead of the header.

Instead, we can replicate something similar to streaming by first fetching only the metadata region at the end of the file, and then making multiple requests for each internal chunk.

### Files are immutable

> Once written, a Parquet file is immutable. No modification or appending can happen to that Parquet file. Instead, create a new Parquet file.

Reviewer 3: For Parquet stored as multi-files, would it be possible to 'append' new rows by adding a new chunk file?

Reviewer 2: Isn't there a main file that would need to be updated to know about the new chunk?

Author: > For Parquet stored as multi-files, would it be possible to 'append' new rows by adding a new chunk file?

Yes, as long as it has the same schema.

> Isn't there a main file that would need to be updated to know about the new chunk?

It depends. There _can_ be a metadata file at the top level in the folder that describes all the Parquet files in the directory. That file doesn't always exist (it's not required by the parquet spec; the parquet spec is mostly concerned with individual files, and people build support for multiple files on top of it).

But it's relatively easy to update that top-level metadata file, compared with creating a new Parquet file.
